public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts

```diff
@@ -41,6 +41,30 @@ function move(node: SyntaxNode, direction: Direction): SyntaxNode | null {
   return node[direction];
 }
 
+/**
+ * Iteratively calls walk with given path until it returns null, then we return the last non-null node.
+ * @param node
+ * @param path
+ */
+function traverse(node: SyntaxNode, path: Path): SyntaxNode | null {
```

Comment:
[enter](https://lezer.codemirror.net/docs/ref/) looks like it would jump to a known position in the child node for a TreeCursor/SyntaxNode, and the parent,firstchild,lastchild,nextSibling references on the nodes are provided as the traversal API from what I can tell. I would argue our implementation of move is expected, from my (limited) understanding @lezer/common is relatively low level abstraction on the syntax tree data type, so the actual traversal implementation is left to the implementers.

Reply:
My bad actually, I was quoting something from memory. I meant [iterate](https://lezer.codemirror.net/docs/ref/#common.Tree.iterate), `enter` is just a property of it.

The reason was inclined to make the comment is that, so far, we've been relying on iterate() for pretty much most of our use cases which you can usually find in [queryUtils](https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/queryUtils.ts#L117).

Reply:
That makes more sense, yeah we could convert the syntaxNodes to tree and use iterate, I think the docs say for larger trees the ideal is to use a treeCursor

> A syntax node provides an immutable pointer to a given node in a tree. When iterating over large amounts of nodes, you may want to use a mutable [cursor](https://lezer.codemirror.net/docs/ref/#common.TreeCursor) instead, which is more efficient.

Which is available on the SyntaxNode via the .cursor() property, which could save us from needing to store references to each node in the tree.

I will create a stub tech/debt issue for us to investigate, but my opinion is that this isn't the PR to address that.

Reply:
https://github.com/grafana/grafana/issues/76143

Reply:
>  large amounts of nodes

I imagine large amount of nodes might mean, for example, a large file in VSCode. I'd love to see some metrics for smaller use cases.

Reply:
Yeah I imagine the overhead for these little QL queries is not much, the current parsing is quite fast, but since it is a blocking activity on every few keystrokes I'd like to keep it that way :P